File: core/tools/allowlist.ts
/**
 * Safe Tool Allowlist & Policy Engine
 * Enforces the rule: AI must NOT receive unrestricted operating-system access.
 * Any future tool execution MUST be explicitly registered in this allowlist with permission boundaries.
 */

export interface ToolDefinition {
  id: string;
  name: string;
  description: string;
  requiresInternet: boolean;
  permissionLevel: 'READ_ONLY' | 'SAFE_EXECUTE' | 'CONFIRMATION_REQUIRED';
  handler?: (params: Record<string, any>) => Promise<any>;
}

export class ToolAllowlist {
  private allowedTools: Map<string, ToolDefinition> = new Map();

  constructor() {
    // Register safe YouTube and music browser tools strictly on allowlisted domain
    this.registerTool({
      id: 'open_youtube',
      name: 'Open YouTube',
      description: 'Safely opens YouTube in the default browser using allowlisted domain https://www.youtube.com',
      requiresInternet: true,
      permissionLevel: 'SAFE_EXECUTE',
    });
    this.registerTool({
      id: 'search_youtube',
      name: 'Search YouTube',
      description: 'Safely searches for songs or videos on YouTube using allowlisted domain https://www.youtube.com',
      requiresInternet: true,
      permissionLevel: 'SAFE_EXECUTE',
    });
    this.registerTool({
      id: 'play_music',
      name: 'Play Music',
      description: 'Safely plays music or opens YouTube music search using allowlisted domain https://www.youtube.com',
      requiresInternet: true,
      permissionLevel: 'SAFE_EXECUTE',
    });
    this.registerTool({
      id: 'stop_music',
      name: 'Stop Music',
      description: 'Safely pauses or stops music playback and media streams without closing unrelated tabs',
      requiresInternet: false,
      permissionLevel: 'SAFE_EXECUTE',
    });
  }

  public registerTool(tool: ToolDefinition): void {
    this.allowedTools.set(tool.id, tool);
  }

  public isAllowed(toolId: string): boolean {
    return this.allowedTools.has(toolId);
  }

  public getTool(toolId: string): ToolDefinition | undefined {
    return this.allowedTools.get(toolId);
  }

  public listTools(): ToolDefinition[] {
    return Array.from(this.allowedTools.values());
  }
}

export const toolAllowlist = new ToolAllowlist();

File: core/tools/allowlist.py
"""
Tool Allowlist & Permission Engine (Python)
Ensures AI cannot execute arbitrary OS commands.
All tools must be explicitly allowlisted and validated.
"""

from typing import Dict, Any, List, Optional
from pydantic import BaseModel


class ToolDefinition(BaseModel):
    id: str
    name: str
    description: str
    requires_internet: bool = False
    permission_level: str = "READ_ONLY"  # READ_ONLY, SAFE_EXECUTE, CONFIRMATION_REQUIRED


class ToolAllowlist:
    def __init__(self):
        self._tools: Dict[str, ToolDefinition] = {}

    def register(self, tool: ToolDefinition) -> None:
        self._tools[tool.id] = tool

    def is_allowed(self, tool_id: str) -> bool:
        return tool_id in self._tools

    def get_tool(self, tool_id: str) -> Optional[ToolDefinition]:
        return self._tools.get(tool_id)

    def list_tools(self) -> List[ToolDefinition]:
        return list(self._tools.values())


tool_allowlist = ToolAllowlist()
